File: 01-Programming-Languages/R/Algorithms/Sorting/merge_sort.R
merge_sort <- function(arr) {
  if (length(arr) <= 1) {
    return(arr)
  }
  
  mid <- length(arr) %/% 2
  left <- merge_sort(arr[1:mid])
  right <- merge_sort(arr[(mid + 1):length(arr)])
  
  merge(left, right)
}

merge <- function(left, right) {
  result <- c()
  i <- 1
  j <- 1
  
  while (i <= length(left) && j <= length(right)) {
    if (left[i] <= right[j]) {
      result <- c(result, left[i])
      i <- i + 1
    } else {
      result <- c(result, right[j])
      j <- j + 1
    }
  }
  
  c(result, left[i:length(left)], right[j:length(right)])
}

arr <- c(38, 27, 43, 3, 9, 82, 10)
cat("Original:", arr, "\n")
cat("Sorted:", merge_sort(arr), "\n")

File: 01-Programming-Languages/Swift/Algorithms/Sorting/merge_sort.swift
func mergeSort(_ arr: [Int]) -> [Int] {
    guard arr.count > 1 else { return arr }
    
    let mid = arr.count / 2
    let left = mergeSort(Array(arr[0..<mid]))
    let right = mergeSort(Array(arr[mid...]))
    
    return merge(left, right)
}

func merge(_ left: [Int], _ right: [Int]) -> [Int] {
    var result: [Int] = []
    var i = 0, j = 0
    
    while i < left.count && j < right.count {
        if left[i] <= right[j] {
            result.append(left[i])
            i += 1
        } else {
            result.append(right[j])
            j += 1
        }
    }
    
    result += left[i...]
    result += right[j...]
    
    return result
}

let arr = [38, 27, 43, 3, 9, 82, 10]
print("Original:", arr)
print("Sorted:", mergeSort(arr))